src/components/Swap/Summary/index.tsx

```diff
@@ -19,6 +22,75 @@ import Summary from './Summary'
 
 export default Summary
 
+enum ReviewState {
+  REVIEWING,
+  ALLOWANCE_PENDING,
+  ALLOWANCE_FAILED,
+  TRADE_CHANGED,
+  SWAP_PENDING,
+}
+
+function useReviewState(
+  trade: InterfaceTrade,
+  onConfirm: () => Promise<void>,
+  allowance: Allowance,
+  doesTradeDiffer: boolean
+) {
+  const [currentState, setCurrentState] = useState(ReviewState.REVIEWING)
+  const { onSubmitSwapClick } = useAtomValue(swapEventHandlersAtom)
+
+  useEffect(() => {
+    if (doesTradeDiffer && currentState === ReviewState.REVIEWING) {
+      setCurrentState(ReviewState.TRADE_CHANGED)
+    } else if (!doesTradeDiffer && currentState === ReviewState.TRADE_CHANGED) {
+      setCurrentState(ReviewState.REVIEWING)
+    }
+  }, [currentState, doesTradeDiffer])
+
+  const triggerSwap = useCallback(async () => {
+    setCurrentState(ReviewState.SWAP_PENDING)
+    onSubmitSwapClick?.(trade)
+    await onConfirm()
+    setCurrentState(ReviewState.REVIEWING)
+  }, [onConfirm, onSubmitSwapClick, trade])
+
+  const prevAllowanceRef = useRef(allowance)
+  // Automatically triggers signing swap tx if allowance requirements are met
+  useEffect(() => {
+    if (prevAllowanceRef.current.state === AllowanceState.REQUIRED && allowance.state === AllowanceState.ALLOWED) {
+      // Prevents swap if trade has updated mid permit2 flow
+      if (doesTradeDiffer) setCurrentState(ReviewState.TRADE_CHANGED)
+      else triggerSwap()
+    }
+    prevAllowanceRef.current = allowance
+  }, [allowance, doesTradeDiffer, triggerSwap])
+
+  const triggerPermit2Flow = useCallback(async () => {
+    if (allowance.state === AllowanceState.REQUIRED) {
+      setCurrentState(ReviewState.ALLOWANCE_PENDING)
+      try {
+        await allowance.approveAndPermit?.()
+      } catch (e) {
+        console.error(e)
+        setCurrentState(ReviewState.ALLOWANCE_FAILED)
+      }
+    }
+  }, [allowance])
+
+  const onStartSwapFlow = useCallback(() => {
+    if (allowance.state === AllowanceState.REQUIRED) {
+      triggerPermit2Flow()
+      // if the user finishes permit2 allowance flow, triggerSwap() is called by useEffect above once state updates
+    } else if (allowance.state === AllowanceState.ALLOWED) {
+      triggerSwap()
+    }
+  }, [allowance, triggerPermit2Flow, triggerSwap])
```

Comment:
okie one other suggestion here:

so we have signing calls being triggered in both callbacks and in `useEffects` - maybe would be best to keep all the logic in one compact `useEffect` like:

```
const [swapFlowStarted, setSwapFlowStarted] = useState(false)

useEffect(() => {
   if (!swapFlowStarted) return 

   if (allowance.state === AllowanceState.REQUIRED) {
      triggerPermit2Flow()
    } else if (allowance.state === AllowanceState.ALLOWED) {
      if (tradeDoesNotDiffer) {
            triggerSwap()
       }
    }
}, [...])
const onStartSwapFlow = () => setSwapFlowStarted(true)
```

could probably get rid of the need to call `triggerSwap` on line 63